src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs

```
@@ -0,0 +1,117 @@
+            }
+
+            bool isManualReset = mode switch
+            {
+                EventResetMode.ManualReset => true,
+                EventResetMode.AutoReset => false,
+                _ => throw new ArgumentOutOfRangeException(nameof(mode), SR.Format(SR.ArgumentOutOfRange_Enum))
+            };
+
+            return CreateEventWaitHandle(initialState, isManualReset, name, out createdNew, eventSecurity);
+        }
+
+        private static unsafe EventWaitHandle CreateEventWaitHandle(bool initialState, bool isManualReset, string name, out bool createdNew, EventWaitHandleSecurity security)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(name));
+            Debug.Assert(security != null);
+
+            EventWaitHandle eventHandle = null;
+
+            fixed (byte* pSecurityDescriptor = security.GetSecurityDescriptorBinaryForm())
+            {
+                var secAttrs = new Interop.Kernel32.SECURITY_ATTRIBUTES
+                {
+                    nLength = (uint)sizeof(Interop.Kernel32.SECURITY_ATTRIBUTES),
+                    lpSecurityDescriptor = (IntPtr)pSecurityDescriptor
+                };
+
+                SafeWaitHandle handle = Interop.Kernel32.CreateEvent(ref secAttrs, isManualReset, initialState, name);
+                ValidateHandle(handle, name, out createdNew);
+
+                AggregateException aggEx = null;
+
+                // Ensure we always close the handle regardless of the OpenExisting call result
+                try
+                {
+                    eventHandle = EventWaitHandle.OpenExisting(name);
+                }
+                catch (Exception ex)
+                {
+                    aggEx = new AggregateException(ex);
+                }
+                finally
+                {
+                    handle.Dispose();
+                }
+
+                if (aggEx != null)
+                {
+                    throw aggEx;
+                }
```

Why wrap in an AggregateException? That usually used to, well, aggregate several exceptions. Would this work?
```c#
                using (handle) EventWaitHandle.OpenExisting(name);
```

---

I'm calling this inside the `fixed` section, and the handle returned by `OpenExisting` needs to be returned by the method.

---

Also, I don't think I should call `TryOpenExisting(string name, out EventWaitHandle handle)` because we need to bubble up the exception to the user.

---

Nothing gets returned when you throw. All you need here is:

``` C#
                try
                {
                    return EventWaitHandle.OpenExisting(name);
                }
                finally
                {
                    // Close our handle as the EventWaitHandle will have it's own.
                    handle.Dispose();
                }
```

---

Which is what @danmosemsft's comment on using does for you. :)